columnar/src/column_values/u128_based/compact_space/build_compact_space.rs

```diff
@@ -93,13 +93,24 @@ pub fn get_compact_space(
         let staged_spaces_sum: u128 = blank_collector.staged_blanks_sum();
         let amplitude_new_compact_space = amplitude_compact_space - staged_spaces_sum;
         let amplitude_new_bits = num_bits(amplitude_new_compact_space);
+
         if amplitude_bits == amplitude_new_bits {
             continue;
         }
         let saved_bits = (amplitude_bits - amplitude_new_bits) as usize * total_num_values as usize;
         // TODO: Maybe calculate exact cost of blanks and run this more expensive computation only,
         // when amplitude_new_bits changes
         let cost = blank_collector.num_staged_blanks() * cost_per_blank;
+
+        // We want to end up with a compact space that fits into 32 bits.
+        // In order to deal with pathological cases, we force the algorithm to keep
+        // refining the compact space the amplitude bits is lower than 32.
+        //
+        // The worst case scenario happens for a large number of u128s regularly
+        // spread over the full u128 space.
+        //
+        // This change will force the algorithm to degenerate into dictionary encoding.
+        // if amplitude_bits > 32 && cost >= saved_bits {
```

Comment:
the change is commented, and that's not how it should work

we could alter the condition, so we always add_blanks the blanks when reaching 32 bits value space.

```rust
let reached_u32_value_space = amplitude_new_bits <= 32 && amplitude_bits > 32;
if  !reached_u32_value_space && cost >= saved_bits {
    continue;
}
```

Reply:
Nice catch! I messed up. In my test I never managed to end up with `amplitude_bits > 32` so I ended up just putting the assert.

Reply:
`test_worst_case_scenario` tests the wrong thing, it calls `add_blanks` unconditionally, but the algorithm could decide that the metadata overhead is too high and not add the blanks to reduce the value space. The logic that needs to be tested in `test_worst_case_scenario` is `get_compact_space`. 10_000 blanks may not be enough

The cost per blank is hardcoded currently `const COST_PER_BLANK_IN_BITS: usize = 36;`

Reply:
oooohh you are absolutely right!